pkg/broker/handler/handler.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"context"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol/pubsub"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors"
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/logging"
+)
+
+// Handler pulls Pubsub messages as events and processes them
+// with chain of processors.
+type Handler struct {
+	// PubsubEvents is the CloudEvents Pubsub protocol to pull
+	// messages as events.
+	PubsubEvents *pubsub.Protocol
+
+	// Processor is the processor to process events.
+	Processor processors.Processor
+
+	// Timeout is the timeout for processing each individual event.
+	Timeout time.Duration
+
+	// cancel is function to stop pulling messages.
+	cancel context.CancelFunc
+}
+
+// Start starts the handler.
+// done func will be called if the pubsub inbound is closed.
+func (h *Handler) Start(ctx context.Context, done func(error)) {
+	ctx, h.cancel = context.WithCancel(ctx)
+	go func() {
+		done(h.PubsubEvents.OpenInbound(ctx))
+	}()
+
+	go h.handle(ctx)
+}
+
+// Stop stops the handlers.
+func (h *Handler) Stop() {
+	h.cancel()
+}
+
+func (h *Handler) handle(ctx context.Context) {
+	for {
```

Comment:
I think we should terminate the loop if ctx is cancelled.

Reply:
I don't think that's necessary. IIUC `Receive` will cancel itself if context is done.

Reply:
It's up to you but I would still recommend doing it.  That way even if the dependencies don't handle context properly, we can terminate as early as we could.

Reply:
Sorry I still don't think it's necessary. Notice `Receive` is a blocking call (it's not a channel). We would need to have a goroutine to call it and pass the message to another channel and do something like

```
for {
  select {
  case msg := <-ch:
    // handle msg
  case <-ctx.Done():
   // stop
  }
}
```

But if you look at the code of cloudevents sdk, it's already what it does. That's for pubsub cancelling.

For processor(s) cancelling, I have commented on the processor interface. A processor shouldn't do any explicit timeout if it can fully rely on its successor(s) for timeout. Unless a processor itself does some long running processing. We should assume processor(s) can handle timeout correctly. It very much simplifies processor code because most of them won't need explicit timeout handling.